chore(ci): ignore private repositories (esp. backstage app, backend) when release plugins

christoph-jerolimov opened the pull request:
## Hey, I just made a Pull Request!

This fixes #1508 by adding this configuration all changeset `config.json`s:

```json
  "privatePackages": {
    "tag": false,
    "version": false
  }
```

See https://github.com/changesets/changesets/issues/1093#issuecomment-2254235349 and https://github.com/changesets/changesets/blob/main/packages/cli/CHANGELOG.md#2251

I tried this locally in the npm package, which contains a changeset and a backstage app.

Without this ignore list it touches these files:

```
	deleted:    .changeset/proud-brooms-clean.md
	modified:   packages/app/package.json
	modified:   plugins/npm/package.json
	new:        packages/app/CHANGELOG.md
	new:        plugins/npm/CHANGELOG.md
```

With this change, it touches only:

```
	deleted:    .changeset/proud-brooms-clean.md
	modified:   plugins/npm/package.json
	new:        plugins/npm/CHANGELOG.md
```

I noticed that the changeset CLI crashes if a workspace contains only private packages and plugins. As discussed, I raise another PR to fix that for the npm plugin.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

04kash reviewed (approved):
lgtm! Should we maybe document this in `CONTRIBUTING.md` that if someone includes an example app/backend they should add `"ignore": ["app", "backend"]` in the  respective workspace's .changeset/config.json?

awanlin reviewed (commented):
Hi @jerolimov, thanks for running with this! I was wondering if would it be possible to leverage the `private` property in the `package.json` to drive this? Packages with `private: true` don't get published which in turn means they don't need to be versioned. Using this would mean that we don't have to manually maintain this list in the `.changeset/config.json`.

04kash commented:
Some of them do have `private: true` but are still being included in the Version Packages PR. Eg: the app package in redhat-argocd has the private property set in [package.json ](https://github.com/backstage/community-plugins/blob/main/workspaces/redhat-argocd/packages/app/package.json#L7) but was still included in this [Version Packages PR](https://github.com/backstage/community-plugins/pull/1530)

jerolimov commented:
Yes, I had the same hope but couldn't find another solution yet.

As said, I also tried first to run yarn changeset version --ignore app --ignore backend, but it works only fine on projects with an backstage app, but  fail on these without.

awanlin reviewed (changes requested):
Hi @jerolimov, I've since found this: https://github.com/changesets/changesets/issues/1093#issuecomment-2254235349. Can we try that instead, please?

If this works we should update this as well: https://github.com/backstage/community-plugins/blob/main/workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json

jerolimov commented:
> Hi @jerolimov, I've since found this: [changesets/changesets#1093 (comment)](https://github.com/changesets/changesets/issues/1093#issuecomment-2254235349). Can we try that instead, please?
> 
> If this works we should update this as well: https://github.com/backstage/community-plugins/blob/main/workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json

Thanks, that's a better approche! I'm looking into this. But after adding

```json
  "privatePackages": {
    "tag": false,
    "version": false
  }
```

with and without updating @changeset/cli it crashes:

```
[nix-shell:~/git/backstage/community-plugins/workspaces/npm]$ yarn changeset
🦋  error TypeError: Cannot read properties of undefined (reading 'packageJson')
🦋  error     at createChangeset (/home/christoph/git/backstage/community-plugins/workspaces/npm/node_modules/@changesets/cli/dist/changesets-cli.cjs.js:416:75)
🦋  error     at add (/home/christoph/git/backstage/community-plugins/workspaces/npm/node_modules/@changesets/cli/dist/changesets-cli.cjs.js:512:26)
🦋  error     at async run (/home/christoph/git/backstage/community-plugins/workspaces/npm/node_modules/@changesets/cli/dist/changesets-cli.cjs.js:1345:5)
```

I'll move this to a draft and keep you updated.

jerolimov commented:
> Hi @jerolimov, I've since found this: [changesets/changesets#1093 (comment)](https://github.com/changesets/changesets/issues/1093#issuecomment-2254235349). Can we try that instead, please?
> 
> If this works we should update this as well: https://github.com/backstage/community-plugins/blob/main/workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json

@awanlin thanks for that suggestion. I changed now all .changeset/config.json and added this configuration:

```
  "privatePackages": {
    "tag": false,
    "version": false
  }
```

It looks like it was expected.

My test case crashed because the npm plugin was marked as private (by myself) and the cli crashes when there is no public package or plugin in a workspace. I will raise an issue or PR on changeset tomorrow.

I removed the `"private": true` flag now from the npm plugin and reverted PR #1506, so a new version PR should be opened for the npm plugin to release 1.0.0. It wasn't published in the origin PR because it was marked as private...

I found just one other private plugin: `@backstage-community/plugin-scaffolder-backend-module-jenkins`, but the jenkins workspace contains multiple other plugins so that the changeset cli shouldn't fail there.

awanlin commented on workspaces/npm/.changeset/proud-brooms-clean.md:
Let's focus this PR on the changes to the `.changeset/config.json` only. This PR should handle the issues with the `npm` plugin itself: [import Grid from '@mui/material/Grid';](https://github.com/backstage/community-plugins/pull/1516)

awanlin replied:
Ugh, not sure why this didn't end up as a link: https://github.com/backstage/community-plugins/pull/1516

awanlin commented on workspaces/npm/packages/app/package.json:
Please revert 👍

awanlin commented on workspaces/npm/plugins/npm/package.json:
Please revert 👍

awanlin reviewed (changes requested):
Awesome work @jerolimov, let's undo the changes for the `npm` plugin as there is a dedicated PR for that. We really want to keep larger PRs like this focused and not leak in other changes 👍 

Also, based on your last comment I'm not sure if you managed to resolve the issue you had or not: https://github.com/backstage/community-plugins/pull/1510#issuecomment-2424264359

jerolimov commented:
> Awesome work @jerolimov, let's undo the changes for the `npm` plugin as there is a dedicated PR for that. We really want to keep larger PRs like this focused and not leak in other changes 👍
> 
> In the PR description can you remove this as well: "This change also reverts #1506 because it doesn't released the npm plugin 1.0.0 because the package was marked as private."
> 
> Also, based on your last comment I'm not sure if you managed to resolve the issue you had or not: [#1510 (comment)](https://github.com/backstage/community-plugins/pull/1510#issuecomment-2424264359)

Done. You're 100% right and I would request the same. This PR now contains just updated `.changeset/config.json` files. :)


OT: The npm release would break until we change it to public, I guess I will open a tiny PR for that in a few minutes...

awanlin reviewed (changes requested):
Hi @christoph-jerolimov, we are still missing the changes to this:  https://github.com/backstage/community-plugins/blob/main/workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json. That will make sure new plugins or migrated plugins have the right settings to start with 👍

christoph-jerolimov commented:
> Hi @christoph-jerolimov, we are still missing the changes to this: https://github.com/backstage/community-plugins/blob/main/workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json. That will make sure new plugins or migrated plugins have the right settings to start with 👍

Hi @awanlin, thanks again.

I can run `yarn changeset` in `workspaces/repo-tools` with this change only when I remove the private attribute from `workspaces/repo-tools/packages/cli/package.json`. The reason is that changelog expects at least one public package in a workspace when using this option.

Adding the config and keeping the private attribute isn't a good option because the next person or job that runs any changeset command will fail.

Removing the private attribute and releasing the cli tool is also a small step to address #1512. (Will not solve it alone for sure.)

Are you fine to have this change as part of this PR?

awanlin commented:
Thanks for the detailed follow up @christoph-jerolimov! Let's go with this:

> 3. Keep the package private and remove the new attributes from workspaces/repo-tools/.changeset/config.json.
> 
> - We can still change the config.json for new workspaces.
> - But the repo-tools is the only workspace with an unchanged config.json itself.

The `repo-tool` workspace is a special case and I think it's fine for now that it's not following this pattern.

christoph-jerolimov commented:
> Thanks for the detailed follow up @christoph-jerolimov! Let's go with this:
> 
> > 3. Keep the package private and remove the new attributes from workspaces/repo-tools/.changeset/config.json.
> > 
> > 
> > * We can still change the config.json for new workspaces.
> > * But the repo-tools is the only workspace with an unchanged config.json itself.
> 
> The `repo-tool` workspace is a special case and I think it's fine for now that it's not following this pattern.

I rebased this PR and applied this change. This mean

1. I removed the change from `workspaces/repo-tools/.changeset/config.json`
2. Kept them in `workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json`
3. Added this to the new workspace `workspaces/mend/.changeset/config.json`

```
grep -E '"updateInternalDependencies": "patch",$' -r workspaces/*/.changeset | wc -l
80

grep -E '"updateInternalDependencies": "patch"$' -r workspaces/*/.changeset        
workspaces/repo-tools/.changeset/config.json:  "updateInternalDependencies": "patch"

grep -E '"updateInternalDependencies": "patch",$' workspaces/repo-tools/packages/cli/src/lib/workspaces/.changeset/config.json 
  "updateInternalDependencies": "patch",
```

Thanks for your patience and let me know if I should address anything else.

christoph-jerolimov commented:
@awanlin can you take a fresh look? Thanks :)

awanlin reviewed (changes requested):
Hi @christoph-jerolimov, something seems to have gone wrong or I misunderstood the direction forward. None of the changes in this PR are using `privatePackages` and they are all still using `ignore`, also there are no `repo-tool` or `mend` changes. Can you take another look, please?

christoph-jerolimov commented:
> Oh what a shame. Sorry, i switched between my pc and notebook and it looks like I rebased the old version. I will take a look later today.

Fixed. (Rebased on the right machine. Code reviews are a life saver.)

awanlin commented on workspaces/repo-tools/.changeset/proud-ducks-knock.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage-community/cli': patch
+---
+
+fix(cli): New workspaces ignores now private repositories when release plugins
```
Not needed, we don't release this package to NPM currently.

christoph-jerolimov replied:
Sure.  I've removed the file and rebased the branch. :crossed_fingers:

christoph-jerolimov commented:
> Super close, if you can just remove the `repo-tools` changeset then we should be good to merge.

Sure. I've removed the file and rebased the branch. 🤞

awanlin reviewed (approved):
Alright, let's get this in place. Thanks again for all your work back and forth on this @christoph-jerolimov 🙏
